feat(event_handler): add custom response validation in OpenAPI utility

amin-farjadi opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number:** #5888

## Summary

Add the option to distinguish between payload validation and response validation.

### Changes

- Optional `has_response_validation_error` argument has been added to `OpenAPIValidationMiddleware`.
- Optional `response_validation_error_http_status` argument has been added to the `ApiGatewayResolver` and its subclasses.
- `ResponseValidationError` exception class has been added to OpenAPI exceptions
- Tests have been added for `APIGatewayRestResolver` response validation.
- Add _Validating responses_ section in the docs

### User experience

Nothing will change. The user will have the option to set a custom response validation http status  for Amazon API Gateway REST and HTTP APIs, Application Load Balancer (ALB), Lambda Function URLs, and VPC Lattice.

```python
from http import HTTPStatus
from typing import Optional

import requests
from pydantic import BaseModel, Field

from aws_lambda_powertools import Logger, Tracer
from aws_lambda_powertools.event_handler import APIGatewayRestResolver
from aws_lambda_powertools.logging import correlation_paths
from aws_lambda_powertools.utilities.typing import LambdaContext

tracer = Tracer()
logger = Logger()
app = APIGatewayRestResolver(
    enable_validation=True,
    response_validation_error_http_code=HTTPStatus.INTERNAL_SERVER_ERROR,  # (1)!
)


class Todo(BaseModel):
    userId: int
    id_: Optional[int] = Field(alias="id", default=None)
    title: str
    completed: bool


@app.get("/todos_bad_response/<todo_id>")
@tracer.capture_method
def get_todo_by_id(todo_id: int) -> Todo:
    todo = requests.get(f"https://jsonplaceholder.typicode.com/todos/{todo_id}")
    todo.raise_for_status()

    return todo.json()["title"]  # (2)!


@logger.inject_lambda_context(correlation_id_path=correlation_paths.API_GATEWAY_HTTP)
@tracer.capture_lambda_handler
def lambda_handler(event: dict, context: LambdaContext) -> dict:
    return app.resolve(event, context)
```


## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

Commit:
feat(openapi-validation): Add response validation flag and distinct exception.

Commit:
feat(api-gateway-resolver): Add option for custom response validation error status code.

Commit:
feat(docs): Added doc for custom response validation error responses.

Commit:
refactor(docs): Make exception handler function name better.

Commit:
feat(unit-test): Add tests for custom response validation error.

Commit:
fix: Formatting.

Commit:
fix(docs): Fix grammar in response validation docs

leandrodamascena commented:
Hey hey @amin-farjadi! Thanks a lot for this PR! I'm working on fixing something else and will review it by Monday, ok?

amin-farjadi commented:
> Hey hey @amin-farjadi! Thanks a lot for this PR! I'm working on fixing something else and will review it by Monday, ok?

Hi @leandrodamascena 👋 Absolutely fine, thanks

leandrodamascena commented:
> > Hey hey @amin-farjadi! Thanks a lot for this PR! I'm working on fixing something else and will review it by Monday, ok?
> 
> Hi @leandrodamascena 👋 Absolutely fine, thanks

I see that our CI is failing with some importing or type annotation. Can you take a look please?

amin-farjadi commented:
Why are the tests checked against python 3.9 when the codebase is not compatible with it? Union operator for typing is used extensively in the codebase (which fails the python 3.9 test).

If I turn `int | None` to `Optional[int]`, formatter will be unhappy. But, happy to do so if needs be

leandrodamascena commented on tests/unit/event_handler/test_response_validation.py:
```diff
@@ -0,0 +1,182 @@
+
+class Todo(BaseModel):
+    userId: int
+    id_: int | None = Field(alias="id", default=None)
```
```suggestion
    id_: Optional[int] = Field(alias="id", default=None)
```

leandrodamascena commented:
> Why are the tests checked against python 3.9 when the codebase is not compatible with it? Union operator for typing is used extensively in the codebase (which fails the python 3.9 test).
> 
> If I turn `int | None` to `Optional[int]`, formatter will be unhappy. But, happy to do so if needs be

There are some inconsistencies when using Pydantic models and Union operations.. Try to make this Optional and remove `future annotations`. I think this can solve the problem for now. If not, I can take a look during the review.

amin-farjadi commented:
I see!

Commit:
fix(unit-test): fix failed CI.

Commit:
bugfix(lint): Ignore lint error FA102, irrelevant for python >=3.9

amin-farjadi commented:
@leandrodamascena added ruff rule `FA102` to be ignored for linting as this project uses python >=3.9 . This resolved the issue I was having in precommit, forcing me to import `annotations` in the test.

Commit:
refactor: make response_validation_error_http_status accept more types and add more detailed error messages.

Commit:
feat(unit-test): add tests for incorrect types and invalid configs

leandrodamascena committed:
Merge branch 'develop' into feature/response-validation

leandrodamascena committed:
Merge branch 'develop' into feature/response-validation

leandrodamascena commented:
I'm starting my review of this PR.

leandrodamascena committed:
Merge branch 'develop' into feature/response-validation

leandrodamascena committed:
Merge branch 'develop' into feature/response-validation

leandrodamascena commented on aws_lambda_powertools/event_handler/api_gateway.py:
```diff
@@ -1515,6 +1520,8 @@ def __init__(
             Each prefix can be a static string or a compiled regex pattern
         enable_validation: bool | None
             Enables validation of the request body against the route schema, by default False.
+        response_validation_error_http_status
+            Enables response validation and sets returned status code if response is not validated.
```
```suggestion
        response_validation_error_http_status
            HTTP status code returned when response validation fails.
```

Response validation is already enabled, we are modifying the customer to inform a specific http status code.

leandrodamascena commented on aws_lambda_powertools/event_handler/api_gateway.py:
```diff
@@ -1496,6 +1500,7 @@ def __init__(
         serializer: Callable[[dict], str] | None = None,
         strip_prefixes: list[str | Pattern] | None = None,
         enable_validation: bool = False,
+        response_validation_error_http_status=None,
```
```suggestion
        response_validation_error_http_code=None,
```

What do you think about this?

amin-farjadi replied:
Looks good

leandrodamascena commented on aws_lambda_powertools/event_handler/api_gateway.py:
```diff
@@ -1530,6 +1537,28 @@ def __init__(
         self._response_builder_class = ResponseBuilder[BaseProxyEvent]
+        self._has_response_validation_error = response_validation_error_http_status is not None
+
+        if response_validation_error_http_status and not enable_validation:
+            msg = "'response_validation_error_http_status' cannot be set when enable_validation is False."
+            raise ValueError(msg)
+
+        if (
+            not isinstance(response_validation_error_http_status, HTTPStatus)
+            and response_validation_error_http_status is not None
+        ):
+
+            try:
+                response_validation_error_http_status = HTTPStatus(response_validation_error_http_status)
+            except ValueError:
+                msg = f"'{response_validation_error_http_status}' must be an integer representing an HTTP status code."
+                raise ValueError(msg) from None
+
+        self._response_validation_error_http_status = (
+            response_validation_error_http_status
+            if response_validation_error_http_status
+            else HTTPStatus.UNPROCESSABLE_ENTITY
+        )
```
```suggestion
        self._response_validation_error_http_status = self._validate_response_validation_error_status(
            response_validation_error_http_status, 
            enable_validation
        )
```

And then we can simplify the logic and move the code for a specific function:

```python
def _validate_response_validation_error_status(self, response_validation_error_http_status, enable_validation):
	        if response_validation_error_http_status and not enable_validation:
	            raise ValueError("'response_validation_error_http_status' cannot be set when enable_validation is False.")
	
	        if response_validation_error_http_status is None:
	            return HTTPStatus.UNPROCESSABLE_ENTITY
	
	        if isinstance(response_validation_error_http_status, HTTPStatus):
	            return response_validation_error_http_status
	
	        try:
	            return HTTPStatus(response_validation_error_http_status)
	        except ValueError:
	            raise ValueError(f"'{response_validation_error_http_status}' must be an integer representing an HTTP status code.")
```

amin-farjadi replied:
Great idea

leandrodamascena commented on aws_lambda_powertools/event_handler/api_gateway.py:
```diff
@@ -2370,6 +2406,25 @@ def _call_exception_handler(self, exp: Exception, route: Route) -> ResponseBuild
+        # OpenAPIValidationMiddleware will only raise ResponseValidationError when
+        # 'self._response_validation_error_http_status' is not None
+        if isinstance(exp, ResponseValidationError):
+            http_status = (
+                self._response_validation_error_http_status
+                if self._response_validation_error_http_status
+                else HTTPStatus.UNPROCESSABLE_ENTITY
+            )
```
```suggestion
            http_status = self._response_validation_error_http_status
```
Am I wrong or aren't we testing this condition before setting the HTTP code? Do we need it again here?

amin-farjadi replied:
Absolutely right

leandrodamascena commented on aws_lambda_powertools/event_handler/middlewares/openapi_validation.py:
```diff
@@ -67,8 +71,14 @@ def __init__(self, validation_serializer: Callable[[Any], str] | None = None):
             Optional serializer to use when serializing the response for validation.
             Use it when you have a custom type that cannot be serialized by the default jsonable_encoder.
+
+        custom_serialize_response_error: ValidationException, optional
+            Optional error type to raise when response to be returned by the endpoint is not
+            serialisable according to field type.
+            Raises RequestValidationError by default.
```
I'm not sure I fully understand this. Could you please explain it to me?

amin-farjadi replied:
I have no idea where this came from :D. Will remove it and add `_has_response_validation_error` param to docstring.

leandrodamascena commented on tests/unit/event_handler/test_response_validation.py:
This shouldn't be a unit test; it's more suited as a functional test. I'd like it moved to https://github.com/aws-powertools/powertools-lambda-python/tree/develop/tests/functional/event_handler/_pydantic.

Also, please avoid adding tests inside classes. Please move the tests to plain functions, and utilizes fixtures when needed.

leandrodamascena commented on ruff.toml:
```diff
@@ -38,7 +38,8 @@ lint.ignore = [
     "PYI024",
-    "A005"
+    "A005",
+    "FA102" # project must require python >= 3.9 making this error obsolete
```
I'll review this once all other reviews are complete.

leandrodamascena reviewed (changes requested):
Hey @amin-farjadi! This is an excellent PR, and I really appreciate your work on this. I've left some comments that I believe we should address before moving to the next round of review.

I'm also curious about potentially enabling this response validation code for individual routes as well. We could have it at both the global level and per route. This isn't a mandatory change, but I'd like to hear your thoughts on this idea.

Let me know what you think, and thanks again for your valuable contribution!

amin-farjadi commented:
Hi @leandrodamascena, thanks for the review. Will implement suggested changes by the end of the week. Regarding custom http status code on a route-level, I will familiarise myself with the `Route` implementation and has a crack at it in this PR.

leandrodamascena commented:
> Hi @leandrodamascena, thanks for the review. Will implement suggested changes by the end of the week. Regarding custom http status code on a route-level, I will familiarise myself with the `Route` implementation and has a crack at it in this PR.

Thanks @amin-farjadi! It's great that you understand the `Route` implementation. However, after further consideration, let's keep this PR as is and create a new issue to discuss the implementation of the route-specific functionality. We can then work on that in a separate PR.

amin-farjadi committed:
refactor: rename response_validation_error_http_status to response_validation_error_http_code

amin-farjadi committed:
refactor(api_gateway): add method for validating response_validation_error_http_code param.

amin-farjadi committed:
fix(api_gateway): fix type and docstring for response_validation_error_http_code param.

amin-farjadi committed:
fix(api_gateway): remove unncessary check of response_validation_error_http_code param being None.

amin-farjadi committed:
fix(openapi-validation): docstring for has_response_validation_error param.

amin-farjadi committed:
refactor(tests): move unit tests into openapi_validation functional test file

leandrodamascena committed:
Merge branch 'develop' into feature/response-validation

amin-farjadi commented:
Hi @leandrodamascena, I believe all your points have been addressed in the recent commits. I have drafted an implemented for route-specific custom response validation http status code which I can add in a subsequent PR

leandrodamascena committed:
Merge branch 'develop' into feature/response-validation

leandrodamascena commented:
> Hi @leandrodamascena, I believe all your points have been addressed in the recent commits. I have drafted an implemented for route-specific custom response validation http status code which I can add in a subsequent PR

Hi @amin-farjadi! Thanks a lot for addressing all the points! I'll review it again today and I'll probably have some very small changes to make, especially in the documentation part, but nothing critical and I can push a commit before approving it, ok?

I'm very happy to have this new feature, it will give customers more flexibility when working with data validation.

Regarding to the validation code per route, please open an issue with a pseudo code showing the experience, this will be super nice!

amin-farjadi commented:
> > Hi @leandrodamascena, I believe all your points have been addressed in the recent commits. I have drafted an implemented for route-specific custom response validation http status code which I can add in a subsequent PR
> 
> Hi @amin-farjadi! Thanks a lot for addressing all the points! I'll review it again today and I'll probably have some very small changes to make, especially in the documentation part, but nothing critical and I can push a commit before approving it, ok?
> 
> I'm very happy to have this new feature, it will give customers more flexibility when working with data validation.
> 
> Regarding to the validation code per route, please open an issue with a pseudo code showing the experience, this will be super nice!

I'm happy to try and close this PR today. Absolutely fine by me, please do feel free.

I drafted this issue for route-specific custom response validation http status code (that's a mouthful!) https://github.com/aws-powertools/powertools-lambda-python/issues/6245

leandrodamascena commented:
Hi @amin-farjadi! Just a quick explanation. A while ago you opened this [issue](https://github.com/aws-powertools/powertools-lambda-python/issues/5887) and we fixed it in this [PR](https://github.com/aws-powertools/powertools-lambda-python/pull/6119). But this created a [regression](https://github.com/aws-powertools/powertools-lambda-python/issues/6216) and I had to revert it while I find the right solution.

So because of this, some tests in the `test_openapi_validation_middleware.py` file are failing, but this is not caused by you. Can you please disable the those tests? Don't remove, just add the decorator `@pytest.mark.skipif(reason="Test temporarily disabled until falsy return is fixed")`?

`test_validation_error_none_returned_non_optional_type`
`test_none_returned_for_falsy_return`
`test_custom_response_validation_error_http_code_invalid_response_none`

Commit:
fix(tests): skipping validation for falsy response

amin-farjadi committed:
Merge branch 'develop' into feature/response-validation

amin-farjadi commented:
> Hi @amin-farjadi! Just a quick explanation. A while ago you opened this [issue](https://github.com/aws-powertools/powertools-lambda-python/issues/5887) and we fixed it in this [PR](https://github.com/aws-powertools/powertools-lambda-python/pull/6119). But this created a [regression](https://github.com/aws-powertools/powertools-lambda-python/issues/6216) and I had to revert it while I find the right solution.
> 
> So because of this, some tests in the `test_openapi_validation_middleware.py` file are failing, but this is not caused by you. Can you please disable the those tests? Don't remove, just add the decorator `@pytest.mark.skipif(reason="Test temporarily disabled until falsy return is fixed")`?
> 
> `test_validation_error_none_returned_non_optional_type` `test_none_returned_for_falsy_return` `test_custom_response_validation_error_http_code_invalid_response_none`

Hi @leandrodamascena, this has been added. Please let me know if further changes are required.

leandrodamascena committed:
Making Ruff happy

leandrodamascena committed:
Refactoring documentation

leandrodamascena reviewed (approved):
Hi @amin-farjadi! Thank you very much for working on all the feedback here! This is a super improvement for this utility!

APPROVED!

amin-farjadi commented:
With pleasure @leandrodamascena 🙌
